### PR TITLE
✅ Add basic test suite

### PR DESCRIPTION
**Pull request by EvolveArt**
Solves #17 

- [x] Adds unit tests on utils functions
- [x] Adds a new `Repository` trait used for mocking
- [x] Basic Integration tests for DB interaction using a `TestContext`
- [x] Basic Integration tests for API Calls

### CODE REVIEW

File: src/handlers/indexers/utils.rs [hunk 1]
```diff
@@ -9,3 +9,70 @@ pub fn get_s3_script_key(id: Uuid) -> String {
 pub fn get_script_tmp_directory(id: Uuid) -> String {
     format!("{}/{}/{}.js", env!("CARGO_MANIFEST_DIR"), "tmp", id)
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (apoorvsadana) [inline src/handlers/indexers/utils.rs:14]: Do these tests really help?

1. We are using the same code inside (which is used inside the function) and in the test case so it's bound to match.
2. If I change the directory later on (which I did in the last PR), I don't really want to change all these test cases as well

I think the directory building doesn't check a very important logic to write test cases for them.

File: src/infra/repositories/indexer_repository.rs [hunk 2]
```diff
@@ -50,73 +52,211 @@ pub struct UpdateIndexerStatusAndProcessIdDb {
...
 }
 
-pub async fn _insert(pool: &Pool<AsyncPgConnection>, new_indexer: NewIndexerDb) -> Result<IndexerModel, InfraError> {
+#[async_trait]
+pub trait Repository {
+    async fn insert(&mut self, new_indexer: NewIndexerDb) -> Result<IndexerModel, InfraError>;
+    async fn get(&self, id: Uuid) -> Result<IndexerModel, InfraError>;
+    async fn get_all(&self, filter: IndexerFilter) -> Result<Vec<IndexerModel>, InfraError>;
+    async fn update_status(&mut self, indexer: UpdateIndexerStatusDb) -> Result<IndexerModel, InfraError>;
+    async fn update_status_and_process_id(
```
Comment (apoorvsadana) [inline src/infra/repositories/indexer_repository.rs:61]: these functions shouldn't be part of the `Repository` trait as they are specific to `IndexerRepository`

File: src/infra/repositories/indexer_repository.rs [hunk 3]
```diff
@@ -50,73 +52,211 @@ pub struct UpdateIndexerStatusAndProcessIdDb {
...
+    fn test_invalid_type_conversion() {
+        let indexer_db = IndexerDb {
+            id: Uuid::new_v4(),
+            status: "Created".to_string(),
+            indexer_type: "InvalidType".to_string(),
+            process_id: Some(1234),
+            target_url: "http://example.com".to_string(),
+        };
+
+        let _: IndexerModel = indexer_db.try_into().unwrap();
```
Comment (apoorvsadana) [inline src/infra/repositories/indexer_repository.rs:260]: can we use test cases here or will that over complicate for now

File: src/tests/common/mod.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+use crate::run_migrations;
+
+// Keep the database info in mind to drop them later
+pub struct TestContext {
+    pub base_url: String,
+    pub db_name: String,
+    pub pool: Pool<AsyncPgConnection>,
+}
+
+impl TestContext {
```
Comment (apoorvsadana) [inline src/tests/common/mod.rs:26]: Why do we need a `TestContext`, can we just create a dummy pool to initialize the repository? Also, how will this work in like Github actions?

**Comment by EvolveArt** [comment]
@apoorvsadana 

More context on this [here](https://snoozetime.github.io/2019/06/16/integration-test-diesel.html). After looking at all the ways possible we could do it. This is the most elegant with the use of `Drop`.

To explain the role of this PR and the different integration tests suites in place :
- `Repository` trait used to mock DB interaction, could directly use DB snapshots instead of mock data but that's fine for now
- `TestContext` is an elegant way of doing `before` `after` hooks within tests.
Before each test, we are creating a new database, run migrations and delete it at the end of the test.
We can see it in action in the `repository` test module where we test each `IndexerRepository` method.
Within the CI it's fairly simple we just spin up a postgres service just like we would locally.

File: src/infra/repositories/indexer_repository.rs (hunk 2 above)
Comment (apoorvsadana) [reply src/infra/repositories/indexer_repository.rs:61]: Hey, just unresolving this one as I still have the doubt, shouldn't `update_status` and `update_status_and_process_id` not be a part of the generic `Repository` trait

File: src/handlers/indexers/create_indexer.rs [hunk 5]
```diff
@@ -32,6 +33,7 @@ async fn build_create_indexer_request(request: &mut Multipart) -> Result<CreateI
     let mut create_indexer_request = CreateIndexerRequest::default();
     while let Some(field) = request.next_field().await.map_err(IndexerError::FailedToReadMultipartField)? {
         let field_name = field.name().ok_or(IndexerError::InternalServerError)?;
+        println!("field_name: {}", field_name);
```
Comment (apoorvsadana) [inline src/handlers/indexers/create_indexer.rs:36]: we can remove this?

File: src/infra/repositories/indexer_repository.rs (hunk 2 above)
Comment (EvolveArt) [reply src/infra/repositories/indexer_repository.rs:61]: #28

**Pull request merged**